examples/generate_bom/README.md

```diff
@@ -42,22 +46,96 @@ options:
   --output_file OUTPUT_FILE
                         The path to the output file. If absent, the CSV will
                         be output to the command line.
-  --attributes_to_extract ATTRIBUTES_TO_EXTRACT
-                        A comma-seperated list of attributes to extract.
 ```
 
 Some examples of how you could run this are:
 
 ```bash
 export ALLSPICE_AUTH_TOKEN= # some token
 
-python3 generate_bom.py "test/test" "test.PrjPcb" # Note that the options are not required!
+python3 generate_bom.py "test/test" "test.PrjPcb" "test.PcbDoc"
+# Note that the options are not required!
 
 python3 generate_bom.py "test/test" "test.PrjPcb" --schdoc_repo "test/test_schdoc" --allspice_hub_url "https://my.selfhosted.example.org" --output_file bom.csv
 
 python3 generate_bom.py "test/test" "test.PrjPcb" --attribute_list "Designator,Manufacturer" # Will only extract these two attributes
 ```
 
+### Customizing the Attributes Extracted by the BOM Script
+
+This script relies on the presence of an `attributes_mapping.json` file next to
+it. This file maps the Component Attributes in the SchDoc files to the columns
+of the BOM. The example version of `attributes_mapping.json` is:
+
+```json
+{
+    "description": ["PART DESCRIPTION"],
+    "designator": ["Designator"],
+    "manufacturer": ["Manufacturer", "MANUFACTURER"],
+    "part_number": ["PART", "MANUFACTURER #"]
+}
+```
+
+In this file, the keys are the names of the columns in the BOM, and the values
+are a list of the names of the attributes in the SchDoc files that should be
+mapped to that column. For example, if your part number is stored either in the
+`PART` or `MANUFACTURER #` attribute, you would add both of those to the list.
+If there is only one attribute, you can omit the list and just use a string. The
+script checks these attributes in order, and uses the _first_ one it finds. So
+if both `PART` and `MANUFACTURER #` are defined, it will use `PART`.
+
+If you need to customize the attributes further or have more complex logic,
+customize the `map_json_to_component` function in the script. Note that the
+AllSpice generated JSON may change at any time, so everything except the
+component attributes may change!
+
 You can also import this python file in another file to use the methods defined
 in it. If you want the BOM in a different format, or have other requirements,
 you can read and adapt this code, or use it as a reference for your own code.
+
+## COGS Calculation
+
+The COGS script depends on the BOM script. It takes a BOM CSV file and computes
+the COGS at various quantities of PCBs made. The COGS is computed by looking up
+the part number using an API you define in the script via the
+`fetch_price_for_part` method. An example is implemented using the cofactr API.
```

Comment:
The problem with having them customize it by changing the function is that there is no real API and no backward compatibility guarantee.  We're providing an example script, and if they fork it, which they're required to do if they need to customize the pricing API, then they're on their own.  It's like putting them in the situation we're in with Gitea.  It _suuuuucks_.

...because when we make changes, it's up to them to incorporate those changes into their forked files.  They likely just won't ever do it unless something breaks, which is a big pain.

One option would be to change this code to be a library that they call.  Then it's just like any library with a SemVer.  The downside is that they're responsible for making their own CLI, meaning every customer needs to redo this work, and they'll all make it different.

Another option that probably better fits in this case is a plugin API where we define an interface using [`Protocol`](https://docs.python.org/3/library/typing.html#typing.Protocol), and they implement it in their own module (i.e. `.py` file).  (Mypy has good [docs on it](https://mypy.readthedocs.io/en/stable/protocols.html).)  The cofactr API code can implement this interface, which serves as both a working default and an example of how to use `Protocol`.

Part of the interface needs to be a version that they specify so that when we make breaking changes, their code can continue to work because we know which version their code expects.  The version can be an integer, which is basically the major version in a SemVer.

The reason for having their customizations in a separate file is so that they can pull in our updates without it conflicting with their changes.  They shouldn't need to touch our files.  Our code can try to import the plugin file and catch the error if it doesn't exist, the way you'd do an optional import in Python.

Reply:
Hm, I see the point. I would *prefer* an ABC based solution over Protocols, if only because the ABC solution will fail noisily if the customer makes a mistake implementing the base class, even if the customer isn't using mypy (which I am almost certain they aren't going to be.) We can version the ABC and be permissive with what we take. Roughly:

```py
class ApiFetcherV1(ABC):
    @abstractmethod
    def fetch(...) -> ...

class BomGenerator():
    api_fetcher: Union[ApiFetcherV1, ApiFetcherV2, ...]

   def run():
       if isinstance(api_fetcher, ApiFetcherV1):
           # treat as v1
       # etc
```

where the user would:

```py
class CofactrApiFetcher(ApiFetcherV1):
   def fetch(...):

# ...

BomGenerator(CofactrApiFetcher()).run()
```

(Pardon any syntax mistakes etc, just sketching it out)

My only concerns here would be in explaining how they're supposed to be using this, as we're expecting this to be implemented by people who may not be software engineers or well-versed in python. The method definition is relatively a very simple concept.

Also, RE: future changes on this script - one of the things I very deliberately tried to do here is to make it as independent of py-allspice (and thereby AllSpice Hub) as possible, so this script is actually much more stable than even the BOM script. It doesn't even use py-allspice as a dependency! This doesn't invalidate the need for a better API solution here, because we might make convenience changes, format changes etc., but it does insulate this script from e.g. JSON structure changes.

Reply:
It's true that we can't expect people will be using a type-checker. But I see little benefit to using an abstract base class in this case.

If they don't use a type-checker, the `Protocol` definition still acts as documentation for the method signatures. We will still type-check _our_ code, and one `Protocol` for each version is what I'd want.

Inheritance is a bad practice, and we don't want to encourage bad habits. Composition and interfaces are generally much better, and duck typing is the dynamic typing way to do that.

With duck typing, when we call a method that's not implemented, that would also raise an exception, so it isn't very different from them not implementing a method in an abstract base class.

Besides the version field check, there's no runtime component. So unlike collections in the standard library, abstract base classes again have no benefit here.

Another difference described by the [PEP about `Protocol`s](https://peps.python.org/pep-0544/#rationale-and-goals) is that they don't need to import the base class. On the other hand, using an abstract base class, we'd need to factor it out to a separate module, have a circular module dependency, or have it all in one file. I think that we'd ideally avoid all of those. Or it at least seems nice to not be required to do those things.

This is what I was thinking for what they would need to implement.

```python
# plugin.py
# No imports needed.

class Plugin:
    version = 1

    def fetch_prices(self, part_number: str) -> dict[int, float]:
        # Their code goes here.
        return {}
```

Nice for not having py-allspice as a dependency 👍 I'm truly scared of creating a support nightmare for ourselves.